### ISSUE TITLE
kill the process group with sudo rights

### ISSUE THREAD

**Issue by ggzengel** [Icinga/icinga2#8723]
## Describe the bug

Icinga2 can't kill checks which are called with sudo.
After IMPI didn't respond the memory was eaten up and the whole system was frozen.

```
[2021-04-16 15:45:51 +0000] warning/Process: Killing process group 17700 ('sudo' '/usr/lib/nagios/plugins/check_ipmi_sensor' '-H' 'localhost' '-L' 'user' '-P' 'pass' '-U' 'icinga2' '-v') after timeout of 60 seconds
[2021-04-16 15:45:51 +0000] warning/Process: Couldn't kill the process group 17700 ('sudo' '/usr/lib/nagios/plugins/check_ipmi_sensor' '-H' 'localhost' '-L' 'user' '-P' 'pass' '-U' 'icinga2' '-v'): [errno 1] Operation not permitted
```

**Comment by julianbrost** [comment]
Which operating system, Icinga version, kernel version and sudo version do you use?

**Comment by ggzengel** [comment]
I think that's a generic problem with sudo and unix, because the icinga user has not the right to kill root processes.

**Comment by ggzengel** [comment]
I found some info for this:
https://serverfault.com/questions/325128/unix-permissions-to-allow-group-users-to-kill-each-others-processes
https://superuser.com/questions/137207/how-to-kill-a-process-started-with-a-different-user-without-being-root-or-sudoer
https://stackoverflow.com/questions/33117159/linux-cap-kill-help-needed
https://stackoverflow.com/questions/13334902/is-it-possible-to-send-a-signal-to-process-that-belongs-to-different-user
https://blog.container-solutions.com/linux-capabilities-in-practice

It looks like Icinga has to execute 'sudo kill' or the installer has to give 'cap_kill' capability to the executable while installing.

There is a 3rd possibility to run Icinga as root and drop rights. But this will change a lot of code.

**Comment by ggzengel** [comment]
For systemd there is:

[Service]
CapabilityBoundingSet=CAP_KILL

**Comment by julianbrost** [comment]
> I think that's a generic problem with sudo and unix, because the icinga user has not the right to kill root processes.

Not necessarily. There are a few exceptions. [`kill(2)`](https://manpages.debian.org/buster/manpages-dev/kill.2.en.html):

> For a process to have permission to send a signal, it must either be privileged (under Linux: have the CAP_KILL capability in the user namespace of the target process), **or the real or effective user ID of the sending process must equal the real or saved set-user-ID of the target process.**

Additionally, sudo forwards signals. [`sudo(8)`](https://manpages.debian.org/buster/sudo/sudo.8.en.html#Signal_handling):

> When the command is run as a child of the sudo process, sudo will relay signals it receives to the command.

So in theory, this could work without an additional sudo invocation or additional privileges, see for example:

```
$ sudo true # cache credentials
[sudo] password for jbrost: 
$ sudo sleep inf &
[1] 140967
$ ps auxw | grep sleep
root      140967  0.1  0.0  17604  7880 pts/4    S    16:07   0:00 sudo sleep inf
root      140971  0.0  0.0   5712   688 pts/4    S    16:07   0:00 sleep inf
jbrost    141008  0.0  0.0   6760  2364 pts/4    S+   16:07   0:00 grep sleep
$ kill 140967
$ ps auxw | grep sleep
jbrost    141068  0.0  0.0   6760  2228 pts/4    S+   16:07   0:00 grep sleep
[1]+  Terminated              sudo sleep inf
```

**Comment by ggzengel** [comment]
>  or the real or effective user ID of the sending process must equal the real or saved set-user-ID of the target process.

that's not given.

> So in theory, this could work without an additional sudo invocation or additional privileges, see for example:

You didn't kill the process group.

**Comment by ggzengel** [comment]
I haven't tested `CapabilityBoundingSet=CAP_KILL` but I expect that it will work.

Do you agree with me that a lot of Icinga instances can't kill checks, because the checks use sudo or SUIDs.
The most checks even timeout by itself and quit.
After years of using check_ipmi it suddenly didn't timeout by itself. Because Icinga killed sudo without the subprocesses it started one more check.
That's worst case and I think Icinga should handle even this case, too.

**Comment by Al2Klimov** [comment]
@julianbrost Do you agree with me that https://github.com/Icinga/icinga2/issues/8723#issuecomment-822626587 is a reasonable workaround?
